tests/test_requests.py

```diff
@@ -315,6 +315,26 @@ def test_headers_on_session_with_None_are_not_sent(self, httpbin):
         prep = ses.prepare_request(req)
         assert 'Accept-Encoding' not in prep.headers
 
+    def test_headers_preserve_order(self, httpbin):
+        """Preserve order when headers provided as OrderedDict."""
+        ses = requests.Session()
+        ses.headers = OrderedDict()
```

Comment:
Common practice is to pass `headers` as `dict`. I have thought that an average user would intuitively use `OrderedDict` in place of `dict`, when he want order to be maintained. `CaseInsensitiveDict` is somewhat internal to requests, so many users may not be even aware of its existence.

But I can change it to `CaseInsensitiveDict` if you want.